### PR TITLE
Added JSON support using JSONPath to extract values

### PR DESCRIPTION
**Pull request by renetruelsen**
I've added JSON support for files and url. Using JSONPath to extract values for datalist editor. Including proper logging :-)

Only JsonDataListSource.cs has been modified. No dependencies.
I haven't updated documentation yet. Waiting to see if the direction is right. 

### Types of changes

- [ ] Documentation change
- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [X] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_

### Checklist

- [X] My code follows the coding style of this project.
- [X] My changes generate no new warnings.
- [X] My change requires a change to the documentation.
- [ ] I have updated the corresponding documentation.
- [X] I have read the **[CONTRIBUTING](CONTRIBUTING)** and **[CODE_OF_CONDUCT](CODE_OF_CONDUCT.md)** documents.

### CODE REVIEW

**Comment by renetruelsen** [comment]
...sorry...I tried to target your existing pull request, but I'm really a rookie when it comes to Git, Forks and pull-requests

**Comment by leekelleher** [comment]
@renetruelsen You've done it correctly! 👍 

The way it works is that you've created this pull-request from your repo's branch to my repo's branch... once I merge that in, it will auto-update pull-request #4 - which is for 'dev/wip/json-datasource' to 'master'.

Overall, it's 2 pull-requests, but that's fine, an extra button click.

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs [hunk 1]
```diff
@@ -3,16 +3,21 @@
...
-using Umbraco.Core;
-using Umbraco.Core.Composing;
 using Umbraco.Core.IO;
 using Umbraco.Core.Logging;
 using Umbraco.Core.PropertyEditors;
 
+// TODO
+// How would you get the string-value from a "key"?
+// This project https://github.com/s3u/JSONPath supports "~" to retrieve keys. However this is not in the original jsonpath-specs.
+// We could implement something similar, which checks the JsonPaths for a ~, and the we'll code-extract the keys. However this is a somewhat shady solution.
```
Comment (leekelleher) [inline src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: I'm happy to stick with whatever JSON.NET supports. I hadn't heard about the "~" for keys - interesting!

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs [hunk 2]
```diff
@@ -24,13 +29,18 @@ public JsonDataListSource(ILogger logger)
...
 
-        public string Description => "Configure the data source to use JSON data.";
+        public string Description => "Configure the data source (file or url) to use JSON data. Data retrieved from urls are set to UTF-8 encoding.";
 
         public string Icon => "icon-brackets";
 
-        public Dictionary<string, object> DefaultValues => default;
+        public Dictionary<string, object> DefaultValues => new Dictionary<string, object>
+        {
+            { "itemsJsonPath", "$..book[?(@.enabled == true)]" },
```
Comment (leekelleher) [inline src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: Let's amend this to a more generic example - would `$.*` work?

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs [hunk 3]
```diff
@@ -56,41 +72,81 @@ public IEnumerable<DataListItem> GetItems()
...
+                        Icon = icon?.ToString() ?? "",
+                        Description = description?.ToString() ?? ""
+                    });
+                }
+
+                if (allNameNull)
+                {
+                    _logger.Error<JsonDataListSource>("Contentment | Logging: No 'names' were found in json data. Consider changing your JSONPath. ");
+                    // We return empty list to emphasize error
+                    items = new List<DataListItem>();
```
Comment (leekelleher) [inline src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: `items` would already be instantiated - or is this to clear out existing items? Could do `items.Clear()`?

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs [hunk 4]
```diff
@@ -56,41 +72,81 @@ public IEnumerable<DataListItem> GetItems()
...
+                    var value = item.SelectToken(ValueJsonPath);
+                    var icon = string.IsNullOrEmpty(IconJsonPath) ? null : item.SelectToken(IconJsonPath);
+                    var description = string.IsNullOrEmpty(DescriptionJsonPath) ? null : item.SelectToken(DescriptionJsonPath);
+
+                    // How should we log if either name or value is empty? Note that empty or missing values are totally legal according to json
+                    if (name == null) _logger.Warn<JsonDataListSource>("Contentment | Logging: No 'name' was found using JSONPath: " + NameJsonPath);
+                    if (value == null) _logger.Warn<JsonDataListSource>("Contentment | Logging: No 'value' was found using JSONPath: " + ValueJsonPath);
+
+                    allNameNull = allNameNull && (name == null); // Will flip first time name is not null
+                    allValueNull = allValueNull && (value == null); // Will flip first time value is not null
```
Comment (leekelleher) [inline src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: Can you clarify your thinking behind tracking if all the names/values are `null`?

With the XML data-source, we check that both the `name` and `value` aren't `null` before adding it to the list.

Comment (renetruelsen) [reply src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: > Can you clarify your thinking behind tracking if all the names/values are `null`?
> 
> With the XML data-source, we check that both the `name` and `value` aren't `null` before adding it to the list.

Yes. I realised during my tests that I had some weird experiences with the dropdown editor. It default-selected the last option, and I couldn't grasp why. Turned out that all my value-fields were blank due to an incorrect jsonpath, which caused this error. Basically it should be ok that some of the name or value fields are empty, technically speaking this is the idea of json; that it's not too strict, but if all names or values are empty, then I would consider that an error and log accordingly.

Makes sense?

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs (hunk 2 above)
Comment (renetruelsen) [reply src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: I wanted to include an example which uses filter expressions, and the book-example refers to the same example that I link to for more information. That's why I chose that.

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs (hunk 3 above)
Comment (renetruelsen) [reply src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: Yes, much cleaner.

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs (hunk 1 above)
Comment (renetruelsen) [reply src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: Yeah, but the ~would be a programmable extension where we need to analyse the jsonpath and act accordingly. A nice feature for the future, so I left it in the comments :)

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs (hunk 2 above)
Comment (leekelleher) [reply src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: OK, let's go with it 👍

File: src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs (hunk 4 above)
Comment (leekelleher) [reply src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs]: OK, that's fine. 👍

**Comment by renetruelsen** [comment]
I have one thing that I am not sure about; I'm not too keen on the allNamesNull and allValuesNull, but they solved a problem.
However - maybe it's better to replace these ideas with: 
 * Any "name" that is null is just logged as a warning and replaced with empty string. No further action needed.  And if any value is null that specific loop/item is skipped (and logged as a warning). 

Personally I think this is more robust and logical that a missing name is visible for the user, but items with a missing value should be discarded. Let me know.

**Comment by renetruelsen** [comment]
...I answered my own question. This solution is better.

**Comment by leekelleher** [comment]
@renetruelsen I'll merge in now. Thanks again for picking up on this one! 🎉

**Comment by renetruelsen** [comment]
Thanks, Lee :-)

**Pull request merged**